File: frontend/src/shared/api/rules.ts
import { api } from "./client"

export interface Rule {
  id: string
  name: string
  priority: number
  conditions_json: {
    keywords_any?: string[]
    description_contains?: string
    normalized_description_equals?: string
    counterparty_contains?: string
    direction?: "in" | "out"
  }
  category_id: string
  is_active: boolean
}

export interface RuleSuggestion {
  id: string
  source_pattern: string
  suggested_conditions_json: Record<string, unknown>
  suggested_category_id: string
  coverage_count: number
  status: string
}

export interface RuleCreate {
  name: string
  priority?: number
  conditions_json: Rule["conditions_json"]
  category_id: string
}

export interface RuleUpdate {
  name?: string
  priority?: number
  conditions_json?: Rule["conditions_json"]
  category_id?: string
  is_active?: boolean
}

export const rulesApi = {
  list: () => api.get<Rule[]>("/rules"),
  create: (body: RuleCreate) => api.post<Rule>("/rules", body),
  update: (id: string, body: RuleUpdate) => api.patch<Rule>(`/rules/${id}`, body),
  delete: (id: string) => api.delete<void>(`/rules/${id}`),

  listSuggestions: () => api.get<RuleSuggestion[]>("/rules/suggestions"),
  acceptSuggestion: (id: string) => api.post<Rule>(`/rules/suggestions/${id}/accept`),
  dismissSuggestion: (id: string) => api.post<void>(`/rules/suggestions/${id}/dismiss`),
}

File: backend/app/api/schemas/rules.py
from pydantic import BaseModel, Field


class RuleRead(BaseModel):
    id: str
    name: str
    priority: int
    conditions_json: dict
    category_id: str
    is_active: bool

    model_config = {"from_attributes": True}


class RuleCreate(BaseModel):
    name: str = Field(..., max_length=255)
    priority: int = 0
    conditions_json: dict = Field(default_factory=dict)
    category_id: str


class RuleUpdate(BaseModel):
    name: str | None = Field(None, max_length=255)
    priority: int | None = None
    conditions_json: dict | None = None
    category_id: str | None = None
    is_active: bool | None = None


class RuleSuggestionRead(BaseModel):
    id: str
    source_pattern: str
    suggested_conditions_json: dict
    suggested_category_id: str
    coverage_count: int
    status: str

    model_config = {"from_attributes": True}
